### DISCUSSION TITLE
Complex DEFAULT values within text() clause possible have bugs

### DISCUSSION THREAD

**Discussion by XCanG** [sqlalchemy/sqlalchemy#9512] [Usage Questions] [answered]
Now, I don't know for sure yet if it bug or not, so I didn't open issue for that, but I will tell what I found during one of my fix.

So, I using new SQLAlchemy 2.0 with new mapped columns for a while, in the model I had this fields:
```py
class User(Base):
    ...
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), nullable=False, init=False)
    updated_at: Mapped[datetime | None] = mapped_column(DateTime, server_default=null(), onupdate=datetime.utcnow, nullable=True, init=False)
```

For some time I didn't find bug here, but eventually I found that my user created_at and updated_at inconsistend and updated_at are behind, so I had cases where created_at are later than updated_at, so the problem was with `func.now()` that was in timezone, but `datetime.utcnow` in UTC (which I need for all DT columns)

I did [quick check](https://stackoverflow.com/a/16610360/3673470) that you can use complex call `now() at time zone 'utc'`, but you need to use parenthesis, so I changed code to:
```py
class User(Base):
    ...
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=text("(now() at time zone 'utc')"), nullable=False, init=False)
    updated_at: Mapped[datetime | None] = mapped_column(DateTime, server_default=null(), onupdate=datetime.utcnow, nullable=True, init=False)
```

However, it kind of work, but in a weird way, in pgAdmin I can see it as:
![image](https://user-images.githubusercontent.com/4011523/226328469-49f1d187-f14d-4f96-9e1f-05c707101822.png)

If you look closely, it wrap statement in `'` and `'utc'` become sort of like SQL injection, I do wonder if I use one `'` it will break SQL.

Also I didn't find any other solution that wouldn't put statement without wrapping it in `'`, so may be you could add method like `raw()`?
Also, besides that, could you add into `func.now()` parameter timezone? so that I can do something like `func.now(timezone='utc')` → `now() at time zone 'utc'`

**Comment by zzzeek** [comment, marked as answer]
hi  -

I can't comment on what pgadmin is illustrating but I can confirm SQLAlchemy is not adding quotes, here's your model in a full example:

```py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import annotations
from sqlalchemy import create_engine
from sqlalchemy import DateTime
from sqlalchemy import null
from sqlalchemy import text
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        server_default=text("(now() at time zone 'utc')"),
        nullable=False,
        init=False,
    )
    updated_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        server_default=null(),
        onupdate=datetime.utcnow,
        nullable=True,
        init=False,
    )


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)
```

SQL emitted, using echo=True we can see it:

```sql
CREATE TABLE "user" (
	id SERIAL NOT NULL, 
	created_at TIMESTAMP WITHOUT TIME ZONE DEFAULT (now() at time zone 'utc') NOT NULL, 
	updated_at TIMESTAMP WITHOUT TIME ZONE DEFAULT NULL, 
	PRIMARY KEY (id)
)


```


psql view:

```sql
$ psql -U scott test
psql (14.3)
Type "help" for help.

test=# \d user;
                                        Table "public.user"
   Column   |            Type             | Collation | Nullable |             Default              
------------+-----------------------------+-----------+----------+----------------------------------
 id         | integer                     |           | not null | nextval('user_id_seq'::regclass)
 created_at | timestamp without time zone |           | not null | (now() AT TIME ZONE 'utc'::text)
 updated_at | timestamp without time zone |           |          | 
Indexes:
    "user_pkey" PRIMARY KEY, btree (id)

```


text() is the "raw" construct you are looking for

**Reply by XCanG** [reply]
Ok, I was runned `\d user` and get:
```sql
 created_at      | timestamp without time zone |           | not null | (now() AT TIME ZONE 'utc'::text)
 updated_at      | timestamp without time zone |           |          |
 ```
 So, looks like it wasn't a bug. I suspected that it was not, because otherwise it would be not be able to create table at all, but I couldn't test it otherwise.

**Reply by CaselIT** [reply]
this seems like a bug of pgAdmin. Maybe it should be reported there?

**Reply by XCanG** [reply]
> this seems like a bug of pgAdmin. Maybe it should reported there?

Yea, good idea!
